### DISCUSSION TITLE
VGA Text Mode (zh-CN)

### DISCUSSION THREAD

**Discussion by giscus** [phil-opp/blog_os#1085] [Post Comments (translated)]
# VGA Text Mode (zh-CN)

VGA 字符模式（VGA text mode）是打印字符到屏幕的一种简单方式。在这篇文章中，为了包装这个模式为一个安全而简单的接口，我们将包装 unsafe 代码到独立的模块。我们还将实现对 Rust 语言格式化宏（formatting macros）的支持。

https://os.phil-opp.com/zh-CN/vga-text-mode/

**Comment by CornerOfSkyline** [comment]
Spin version in the article used "0.4.9", but the feature "const_fn" has been removed in current rust version. We can change the spin version to "0.9.2" to solve the problem.

**Comment by toksdotdev** [comment]
Thanks a lot for the article. Was really helpful.

For someone who'll like to use the most recent version of `volatile` (`0.4.4` as at the time of writing), you'll need to implement the `Deref` and `DerefMut` trait for `ScreenChar`. 

Something along the lines of:
```rust

impl DerefMut for ScreenChar {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self
    }
}

impl Deref for ScreenChar {
    type Target = ScreenChar;

    fn deref(&self) -> &Self::Target {
        self
    }
}

```

Asides from that, everything in the article pretty much remains the same.

**Comment by TisnKu** [comment]
Why `ascii_code` is before `color_code` while VGA has color code on higher bytes?

Is it because of little endian?

**Comment by romatthe** [comment]
Since we're using nightly Rust here, is there a reason why we can't use `core::ptr::write_volatile` and `core::ptr::read_volatile` here instead of using the `volatile` crate? Or is the article simply saying it's better to use an ergonomic crate like `volatile` here instead of directly using unsafe code ourselves? Either one is fine by me, but we're going to need to learn how to built our own safe abstractions in this domain anyway.

**Comment by bjmayor** [comment]
赞

**Comment by llllishuo** [comment]
I don't know why this problem occurred
When I run ```cargo bootimage``` The error manifests as
```
error[E0463]: can't find crate for `std`
  |
  = note: the `x86_64-unknown-none` target may not support the standard library
  = note: `std` is required by `os_rust_demo` because it does not declare `#![no_std]`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

error: cannot find attribute `derive` in this scope
  --> src/vga_buffer.rs:35:3
   |
35 | #[derive(Debug, Clone, Copy, PartialEq, Eq)]
   |   ^^^^^^
   |
help: consider importing this attribute macro
   |
3  + use core::prelude::rust_2024::derive;
   |
```
in  .cargo/config.toml
```
$ cat .cargo/config.toml
[unstable]
build-std = ["core", "compiler_builtins"]
build-std-features = ["compiler-builtins-mem"]


[build]
target = "x86_64-unknown-none.json"


[target.'cfg(target_os = "none")']
runner = "bootimage runner"
```

**Reply by llllishuo** [reply]
I have discovered my problem.  I created a new lib.rs.  Perhaps this is the reason that caused the  no_std  to fail

**Reply by xuanplus** [reply]
Maybe you can try the edition-3, which has a better way to set target and so on.

**Reply by phil-opp** [reply]
> I have discovered my problem. I created a new lib.rs. Perhaps this is the reason that caused the no_std to fail

Yes, that seems likely. Every lib.rs or main.rs file that should be `no_std` needs the `#![no_std]` attribute.

**Comment by leolin-idah** [comment]
It's wonderful

**Comment by image1005** [comment]
我使用的是目前最新版本的Rust，添加格式化宏时，需要添加use core::fmt::{self, Write};才可以运行
